airflow/providers/google/cloud/operators/dataflow.py

```diff
@@ -324,6 +344,23 @@ class DataflowTemplatedJobStartOperator(BaseOperator):
             `https://cloud.google.com/dataflow/pipelines/specifying-exec-params
             <https://cloud.google.com/dataflow/docs/reference/rest/v1b3/RuntimeEnvironment>`__
     :type environment: Optional[dict]
+    :param wait_until_finished: (Optional)
+        If True, wait for the end of pipeline execution before exiting. If False,
+        it only waits for it to starts (``JOB_STATE_RUNNING``).
+
+        The default behavior depends on the type of pipeline:
+
+        * for the streaming pipeline, wait for jobs to start,
+        * for the batch pipeline, wait for the jobs to complete.
+
+        .. warning::
+
+            You cannot call ``PipelineResult.wait_until_finish`` method in your pipeline code for the operator
```

Comment:
Note that `wait_until_finish` is called implicitly by `with Pipeline() as p:` as well. Since `with` is the recommended way to run pipelines, I am hesitant about this part. Especially since the user is likely to have not even written the template's pipeline code themselves.

Reply:
Partially I agree, but partially not.
Default behaviour is:
 -  for the streaming pipeline, wait for jobs to start,
  - for the batch pipeline, wait for the jobs to complete.

But there may be the specific cases like:

- user doesn't want to wait for the end of the batch job and knows for sure that templated batch job not `wait_until_finish`
- user want to wait until streaming job will be cancelled/drained (e.g by some external api call or web UI)

It will give possibility to conscious dataflow users for more flexible DAGs if needed. What do you think about it?

Reply:
I am still confused about this part: `You cannot call ``PipelineResult.wait_until_finish`` method in your pipeline code for the operator to work properly.` What exactly will happen if `pipeline.wait_until_finish()` is called and the `wait_until_finished` param is set?

Reply:
The process of starting the Datafłow job in Airflow consists of two:
- running a subprocess and reading the stderr/stderr log for the job id.
- loop waiting for the end of the job ID from the previous step. This loop checks the status of the job.

Step two is started just after step one has finished, so if you have `wait_until_finished` in your pipeline code, step two will not start until the process stops. When this process stops, steps two will run, but it will only execute one iteration as the job will be in a terminal state.

If you in your pipeline do not call the `wait_for_pipeline` method but pass `wait_until_finish =True` to the operator,  the second loop will wait for the job's terminal state.

If you in your pipeline do not call the `wait_for_pipeline` method, and you pass `wait_until_finish =False` to the operator, the second loop will wait for the running state only.

Reply:
Thanks for the explanation @mik-laj. The behaviors you describe seem acceptable.

@TobKed Can you include details like Kamil's comment in the pydocs?

Reply:
Thanks @mik-laj 
@ibzib sure, just pushed fixup with improved docs :)